src/core/sorted_map.h

```diff
@@ -92,8 +101,248 @@ class SortedMap {
                absl::FunctionRef<bool(sds, double)> cb) const;
 
  private:
-  dict* dict_ = nullptr;
-  zskiplist* zsl_ = nullptr;
+  struct RdImpl {
+    struct dict* dict = nullptr;
+    zskiplist* zsl = nullptr;
+  };
+
+  struct DfImpl {
+    ScoreMap* score_map = nullptr;
+    BPTree<uint64_t>* bptree = nullptr;  // just a stub for now.
+  };
+
+  struct SizeOp {
```

Comment:
The couple of hundred lines below are a little ridiculous

Using virtuality for the interface would've saved you all of them (while you can still store the two versions inline inside a variant)

Reply:
https://stackoverflow.com/questions/57726401/stdvariant-vs-inheritance-vs-other-ways-performance#:~:text=the%20outcome%20much)-,Update%20(6),-I%20made%20some

Reply:
I am not a fan of this boilerplate but I did not understand the suggestion. Please note that some operators are readers, and some are writers.
What are you suggesting exactly?

Reply:
I immediately thought about this

```c++
struct TreeInterface {
    virtual void Add() = 0;
};

struct RedisTree : public TreeInterface {
    void Add() override {}
};

struct DflyTree : public TreeInterface {
    void Add() override {}
};

struct TreeWrapper {
    TreeInterface *operator->() {
        return visit([](auto& impl) -> TreeInterface* { return &impl; }, impl);
    }

    variant<RedisTree, DflyTree> impl;
};
```

Reply:
Actually its double dispatch because we call visit to get the pointer 💀 It can just be stored separately I guess

Reply:
Maybe at least having the implementations separately instead of all the structs is more readable, i.e.

```c++
// impls from above

struct TreeWrapper {
  void Add() {
        visit([](auto& impl) { return impl.Add(); }, impl);
    }
}
```

Reply:
This also does the job

```c++


#define WRAPFUNC(rettype, signature) rettype signature { visit([&](auto& impl) -> rettype { return impl.signature; }, impl); }

struct TreeWrapper {
    WRAPFUNC(void, Add());

    variant<RedisTree, DflyTree> impl;
};
```